File: sennaar-rs/src/registry/rawtype.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::cpl::CExpr;
use crate::Identifier;


#[derive(Debug, Clone, Serialize, Deserialize)]
#[derive(JsonSchema)]
#[serde(tag = "$kind")]
pub enum Type<'a> {
    IdentifierType(Box<IdentifierType>),
    ArrayType(Box<ArrayType<'a>>),
    PointerType(Box<PointerType<'a>>)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[derive(JsonSchema)]
pub struct IdentifierType {
    pub ident: Identifier
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[derive(JsonSchema)]
pub struct ArrayType<'a> {
    pub element: Type<'a>,
    pub length: CExpr<'a>
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[derive(JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct PointerType<'a> {
    pub pointee: Type<'a>,
    pub is_const: bool,
    pub pointer_to_one: bool
}

File: sennaar-kt/src/main/kotlin/registry/rawtype.kt
package club.doki7.sennaar.registry

import club.doki7.sennaar.Identifier
import club.doki7.sennaar.cpl.CExpr
import kotlinx.serialization.Serializable

@Serializable
sealed interface Type

@Serializable
data class IdentifierType(var ident: Identifier) : Type {
    constructor(ident: String) : this(Identifier(ident))
}

@Serializable
data class ArrayType(var element: Type, var length: CExpr?) : Type

@Serializable
data class PointerType(
    var pointee: Type,
    var isConst: Boolean,
    var pointerToOne: Boolean,
    var nullable: Boolean
) : Type
